internal/controller/apiextensions/composite/composition_ptf.go

```diff
@@ -0,0 +1,993 @@
+	sort.Slice(refs, func(i, j int) bool {
+		ri, rj := refs[i], refs[j]
+		return ri.APIVersion+ri.Kind+ri.Name < rj.APIVersion+rj.Kind+rj.Name
+	})
+
+	s.Composite.SetResourceReferences(refs)
+}
+
+// A ComposedResourceObserverChain runs a slice of ComposedResourceObservers.
+type ComposedResourceObserverChain []ComposedResourceObserver
+
+// ObserveComposedResources runs the slice of ComposedResourceObservers.
+func (o ComposedResourceObserverChain) ObserveComposedResources(ctx context.Context, s *PTFCompositionState) error {
+	for _, cro := range o {
+		if err := cro.ObserveComposedResources(ctx, s); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// A ReadinessObserver observes composed resource state and updates it to
+// indicate whether each composed resource is ready per the readiness checks
+// associated with each resource, which are derived from their P&T resource
+// template and/or Composition Function desired state.
+type ReadinessObserver struct {
+	check ReadinessChecker
+}
+
+// NewReadinessObserver returns a ComposedResourceObserver that observes whether
+// composed resources are ready.
+func NewReadinessObserver(c ReadinessChecker) *ReadinessObserver {
+	return &ReadinessObserver{check: c}
+}
+
+// ObserveComposedResources to determine their readiness.
+func (o *ReadinessObserver) ObserveComposedResources(ctx context.Context, s *PTFCompositionState) error {
+	for _, cd := range s.ComposedResources {
+		rcfgs := append(ReadinessChecksFromTemplate(cd.Template), ReadinessChecksFromDesired(cd.Desired)...)
+		ready, err := o.check.IsReady(ctx, cd.Resource, rcfgs...)
+		if err != nil {
+			return errors.Wrapf(err, errFmtReadiness, cd.ResourceName, cd.Resource.GetObjectKind().GroupVersionKind().Kind, cd.Resource.GetName())
+		}
+
+		s.ComposedResources.Merge(ComposedResourceState{
+			ComposedResource: ComposedResource{
+				ResourceName: cd.ResourceName,
+				Ready:        ready,
+			},
+		})
+	}
+
+	return nil
+}
+
+// A ConnectionDetailsObserver extracts XR connection details from composed
+// resource state. The details to extract are derived from each composed
+// resource's P&T resource template and/or Composition Function desired state.
+type ConnectionDetailsObserver struct {
+	details ConnectionDetailsExtractor
+}
+
+// NewConnectionDetailsObserver returns a ComposedResourceObserver that observes
+// composed resources in order to extract XR connection details.
+func NewConnectionDetailsObserver(e ConnectionDetailsExtractor) *ConnectionDetailsObserver {
+	return &ConnectionDetailsObserver{details: e}
+}
+
+// ObserveComposedResources to extract XR connection details.
+func (o *ConnectionDetailsObserver) ObserveComposedResources(ctx context.Context, s *PTFCompositionState) error {
+	for _, cd := range s.ComposedResources {
+		ecfgs := append(ExtractConfigsFromTemplate(cd.Template), ExtractConfigsFromDesired(cd.Desired)...)
+		e, err := o.details.ExtractConnection(cd.Resource, cd.ConnectionDetails, ecfgs...)
```

Comment:
I am trying to set some connection secret keys using `FromFieldPath` and noticed that what we have as `cd.Resource` is what we desired rather than the observed state. The problem with this is, I cannot use a path that is filled by the controllers, e.g. atProvider, or resolved references.

Reply:
That should work on a subsequent reconcile though, right? Desired resources should be a superset of observed resources, because we start by loading up (existing) observed resources, then run composition to get our desired state.

I'm imagining a flow like:

1. Reconcile 1 composes the resources and applies them.
2. Reconcile 1 tries to extract connection details from a field path (or connection secret key) that doesn't yet exist.
3. Reconcile 1 finishes.
4. Reconcile 2 gets the resources composed by reconcile 1, which should now have status, conn details etc.
5. Reconcile 2 composes again, probably mostly no-ops this time.
6. Reconcile 2 tries to extract connection details, which are now there.

Am I missing something? Is the observer failing and returning an error, or never returning the connection details you're asking for?

Reply:
I tried with latest changes today and observed a completely different behavior that [FromFieldPath](https://github.com/negz/xfns/pull/1/files#diff-0d0c4c0d411d737742c3cce1e489ae1ed30923b56e4ded9217fb2a081652a30eR20) worked well but [FromConnectionDetailKey](https://github.com/negz/xfns/pull/1/files#diff-0d0c4c0d411d737742c3cce1e489ae1ed30923b56e4ded9217fb2a081652a30eR24) didn't. 

Unfortunately, I didn't have time to spent more time to dig this in but writing here just in case you have any idea why that could happen.

Reply:
@turkenh I think this is because I renamed `FromConnectionDetailKey` back to `FromConnectionSecretKey` per your suggestion in another comment thread -https://github.com/crossplane/crossplane/pull/3465#discussion_r1081968598.